src/compile.rs

```diff
@@ -320,6 +324,205 @@ pub fn validate_functions<'a, 'b>(
     errors.into_result(())
 }
 
+pub fn validate_modules(policies: &Vec<PolicyFile>, types: &TypeMap) -> Result<(), CascadeErrors> {
+    let mut errors = CascadeErrors::new();
+
+    // Store all modules across files in a vector
+    let mut modules_vec: Vec<(&SimpleFile<String, String>, &Module)> = Vec::new();
+    for p in policies {
+        for e in &p.policy.exprs {
+            let d = match e {
+                Expression::Decl(d) => d,
+                _ => continue,
+            };
+            match d {
+                Declaration::Type(_) => continue,
+                Declaration::Func(_) => continue,
+                Declaration::Mod(m) => {
+                    modules_vec.push((&p.file, m));
+                }
+            };
+        }
+    }
+
+    // Validate that module contents exist
+    for (file, module) in &modules_vec {
+        for domain in &module.domains {
+            match types.get(domain.as_ref()) {
```

Comment:
The match blocks for domain and resource are nearly identical, and quite long.  I think they need to get pulled out into a function or macro to reduce code duplication.